core/storage/statistics/gae_models.py

```diff
@@ -737,3 +845,247 @@ def _get_entity_id(cls, exploration_id, exploration_version, state_name,
+    @classmethod
+    def insert_state_rule_answer_log_entity(cls, item):
+        """Attemps to instance a StateRuleAnswerLogModel entity, splitting up
+        its entities as necessary. This method guarantees more than one
+        LargeAnswerBucketModel entity will be written, otherwise the entity will
+        not be split up.
+        """
+        total_answer_count = len(item.answers)
+        if total_answer_count <= cls._MAX_ANSWERS_PER_BUCKET:
+            raise Exception(
+                'Cannot split up entity with less than max answers: %s' % (
+                    item.id))
+        full_bucket_count = total_answer_count / cls._MAX_ANSWERS_PER_BUCKET
+        remaining_answer_count = (
+            total_answer_count % cls._MAX_ANSWERS_PER_BUCKET)
+        answer_list = item.answers.items()
+        # TODO(bhenning): Figure out a better way to deal with this situation.
+        # This sort is only here for the benefit of tests, but it might be slow
+        # enough to cause the transaction to fail.
+        answer_list.sort(key=operator.itemgetter(0))
+        for i in xrange(full_bucket_count):
+            bucket_id = cls.get_new_id('')
+            start_shard_answer_index = i * cls._MAX_ANSWERS_PER_BUCKET
+            end_shard_answer_index = (i + 1) * cls._MAX_ANSWERS_PER_BUCKET
+            sharded_answer_list = answer_list[
+                start_shard_answer_index:end_shard_answer_index]
+            bucket_entity = cls(
+                id=bucket_id,
+                log_model_item_id=item.id,
+                answers=dict(sharded_answer_list),
+                log_model_last_update=item.last_updated)
+            bucket_entity.put()
+
+        if remaining_answer_count > 0:
+            bucket_id = cls.get_new_id('')
+            sharded_answer_list = answer_list[
+                full_bucket_count*cls._MAX_ANSWERS_PER_BUCKET:]
+            bucket_entity = cls(
+                id=bucket_id,
+                log_model_item_id=item.id,
+                answers=dict(sharded_answer_list),
+                log_model_last_update=item.last_updated)
+            bucket_entity.put()
+
+
+class MigratedAnswerModel(base_models.BaseModel):
+    """A temporary model which maps answers in StateRuleAnswerLogModel and
+    LargeAnswerBucketModel to answers migrated and inserted in
+    StateAnswersModel. This model can be used to verify a given answer exists
+    (and the correct number of times) in StateAnswersModel. Although the
+    AnswerMigrationJob is not idempotent, this model aims to help make it closer
+    to idempotent.
+    """
+    # TODO(bhenning): Remove this model once the AnswerMigrationJob has
+    # successfully run in production.
+
+    exploration_id = ndb.StringProperty(indexed=True, required=True)
+    state_name = ndb.StringProperty(indexed=True, required=True)
+    exploration_versions = ndb.IntegerProperty(indexed=False, repeated=True)
+    started_migration = ndb.BooleanProperty(indexed=False, default=False)
```

Comment:
Consider adding comments to each of these properties. I'm not sure what some of them mean.